File: BionicLimbSimulation_unity/Assets/Scripts/Classification/RuleBasedClassifier.cs
using System;
using System.Collections.Generic;
using UnityEngine;

namespace BionicLimb.Classification
{
    /// <summary>
    /// Rule-based fallback classifier using channel RMS thresholds.
    /// Spec §7.4
    /// </summary>
    public class RuleBasedClassifier : IGestureClassifier
    {
        private static readonly string[] Names =
            { "rest", "fist", "open_hand", "pinch", "point", "thumbs_up", "peace", "ok" };

        public string[] GestureNames => Names;

        // Threshold rules: (primary condition, secondary condition)
        // Each entry: Func<float[], bool> for primary, Func<float[], bool> for secondary
        private static readonly (Func<float[], bool> primary, Func<float[], bool> secondary)[] Rules =
        {
            // 0 rest
            (ch => ch[0]<0.15f && ch[1]<0.15f && ch[2]<0.15f && ch[3]<0.15f &&
                   ch[4]<0.15f && ch[5]<0.15f && ch[6]<0.15f && ch[7]<0.15f, _ => true),
            // 1 fist
            (ch => ch[0]>0.4f && ch[1]>0.35f, ch => ch[2]<0.2f),
            // 2 open_hand
            (ch => ch[2]>0.4f && ch[3]>0.35f, ch => ch[0]<0.2f),
            // 3 pinch
            (ch => ch[4]>0.4f && ch[0]>0.3f, ch => ch[2]<0.25f),
            // 4 point
            (ch => ch[2]>0.35f && ch[0]>0.3f, ch => ch[4]<0.2f),
            // 5 thumbs_up
            (ch => ch[4]>0.45f && ch[5]>0.3f, ch => ch[0]<0.2f),
            // 6 peace
            (ch => ch[2]>0.4f && ch[0]>0.25f, ch => ch[4]<0.2f),
            // 7 ok
            (ch => ch[4]>0.4f && ch[0]>0.3f, ch => ch[2]>0.2f),
        };

        public GestureResult Classify(float[,] window)
        {
            float[] rms = ComputeWindowRms(window);

            // Find first matching rule (rest checked last after others)
            int winner = 0; // default rest
            for (int g = 1; g < Rules.Length; g++)
            {
                if (Rules[g].primary(rms) && Rules[g].secondary(rms))
                {
                    winner = g;
                    break;
                }
            }
            // Explicitly check rest as override if all channels low
            if (Rules[0].primary(rms)) winner = 0;

            // Confidence = ratio of rule-matching channels
            float confidence = ComputeConfidence(rms, winner);

            float[] probs = new float[Names.Length];
            probs[winner] = confidence;
            float remaining = (1f - confidence) / (Names.Length - 1);
            for (int i = 0; i < Names.Length; i++)
                if (i != winner) probs[i] = remaining;

            return new GestureResult
            {
                GestureId = winner,
                GestureName = Names[winner],
                Confidence = confidence,
                AllProbabilities = probs
            };
        }

        private static float[] ComputeWindowRms(float[,] window)
        {
            int n = window.GetLength(0);
            float[] rms = new float[8];
            for (int c = 0; c < 8; c++)
            {
                float sum = 0f;
                for (int i = 0; i < n; i++) sum += window[i, c] * window[i, c];
                rms[c] = Mathf.Sqrt(sum / n);
            }
            return rms;
        }

        private static float ComputeConfidence(float[] rms, int gestureId)
        {
            if (gestureId == 0) // rest: confidence based on how low all channels are
            {
                float maxVal = 0f;
                foreach (var v in rms) maxVal = Mathf.Max(maxVal, v);
                return Mathf.Clamp01(1f - maxVal / 0.15f);
            }
            // Count channels that are "active" (above 0.1) and compare to expected
            int activeCount = 0;
            for (int c = 0; c < 8; c++) if (rms[c] > 0.1f) activeCount++;
            return Mathf.Clamp01(activeCount / 4f); // normalised to rough expected active count
        }

        public void Dispose() { }
    }
}

File: bionic-limb-threejs/src/classification/RuleBasedClassifier.ts
import type { IGestureClassifier, GestureResult } from './IGestureClassifier.js';
import { GesturePoseLibrary } from '../hand/GesturePoseLibrary.js';

const THRESHOLD = 0.3;

export class RuleBasedClassifier implements IGestureClassifier {
  classify(features: Float32Array): GestureResult {
    const active = (ch: number) => features[ch] > THRESHOLD;
    const activeCount = Array.from({ length: 8 }, (_, i) => features[i] > THRESHOLD ? 1 : 0)
      .reduce((a: number, b: number) => a + b, 0);

    let gestureId: number;

    if (activeCount >= 5) {
      gestureId = 1; // Fist
    } else if (active(0) && active(1) && active(4)) {
      gestureId = 7; // Victory
    } else if (active(4) && active(5)) {
      gestureId = 6; // ThumbUp
    } else if (active(0) && !active(4)) {
      gestureId = 5; // Point
    } else if (active(1) && active(2)) {
      gestureId = 4; // PinchMiddle
    } else if (active(0) && active(1)) {
      gestureId = 3; // PinchIndex
    } else if (activeCount > 0 && activeCount <= 1) {
      gestureId = 2; // Open
    } else {
      gestureId = 0; // Rest
    }

    const confidence = activeCount === 0 ? 0 : activeCount / 8;
    const gestureName = GesturePoseLibrary.gestureNames[gestureId];

    const allProbabilities = new Float32Array(8);
    if (activeCount === 0) {
      allProbabilities[0] = 1.0;
    } else {
      allProbabilities[gestureId] = confidence;
      const remaining = (1 - confidence) / 7;
      for (let i = 0; i < 8; i++) {
        if (i !== gestureId) allProbabilities[i] = remaining;
      }
    }

    return { gestureId, gestureName, confidence, allProbabilities };
  }

  dispose(): void {}
}
